accepted/0000-plug-an-play.md

```diff
@@ -0,0 +1,302 @@
+
+Adapting the testsuite to a different manager is a matter of implementing an adapter matching the package manager you want to test, then enabling the specs you want to validate. You can take a look at [the Yarn adapter](https://github.com/yarnpkg/yarn/blob/master/packages/pkg-tests/yarn.test.js) for an example on how we implemented this for our package manager.
+
+### B. Plug'n'Play api
+
+```
+interface {
+
+  VERSIONS: {std: 1, [extension]: number};
+
+  findPackageLocator(path: string): {name: null, reference: null};
+  findPackageLocator(path: string): {name: string, reference: string};
+
+  getPackageInformation({name, reference}: {name: null, reference: null}): {packageLocation: string, packageDependencies: Map<string, string>, [key: string]: any};
+  getPackageInformation({name, reference}: {name: string, reference: string}): {packageLocation: string, packageDependencies: Map<string, string>, [key: string]: any};
+
+  resolveToUnqualified(request: string, issuer: string): string;
+  resolveUnqualified(unqualified: string, {extensions?: Array<string>}): string;
+  resolveRequest(request: string, issuer: string): string;
+
+  setup(void): void;
+
+}
+```
+
+### C. formal plug'n'play guarantees
+
+* A package **MUST** be able to get the value exported by the main entries of its dependencies using the `require` function, or through an import statement.
+* A package **MUST** be able to get the filesystem path to a file stored in one of its dependencies by using the `require.resolve` function.
+* A package listing a peer dependency **MUST** obtain the exact same instance of this peer dependency when using `require` than its immediate parent in the dependency tree would. This process is applied recursively.
+
+* A package **MUST NOT** be able to require a package that isn't listed in its dependency detail. The dependency detail is the sum of `dependencies` and `peerDependencies` for all packages, plus `devDependencies` for the top level package if running in development mode.
+    * An exception is made if the package being required is listed in the dependency detail of the top-level. In this case, the package making the request will obtain the exact same **instance** than if the top-level package had made the require call (note the emphasis on instance rather than version).
+    * We however discourage packages from relying on this exception, since it's only been implemented to lower the adoption cost and help plugin systems. Packages should prefer using `peerDependencies` if applicable.
+* Two packages depending on the same reference of the same dependency that itself has a transitive peer dependency **MUST** get the exact same instance of this dependency, whatever their locations in the dependency tree are.
+* Two packages depending on the same reference of the same dependency that itself has a transitive peer dependency **MUST** get different instances of this dependency.

```

Comment:
This rule has the same condition as the previous one, but have different results. Is it a mistook?

Reply:
Yes, the first one should be:

> * Two packages depending on the same reference of the same dependency that *doesn't* have any transitive peer dependencies **MUST** get the exact same instance of this dependency, whatever their locations in the dependency tree are.

In short, if two packages resolve to `lodash 1.0.0`, they are guaranteed to share the exact same 
instance (because `lodash` doesn't have any peer dependency).

If two packages depend on `react-dom 1.0.0`, they are guaranteed to each have their own instance (because `react-dom` has a peer dependency on `react`).